storage/store_pool.go

```diff
@@ -0,0 +1,298 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Bram Gruneir (bram+code@cockroachlabs.com)
+
+package storage
+
+import (
+	"container/heap"
+	"sort"
+	"sync"
+	"time"
+
+	gogoproto "github.com/gogo/protobuf/proto"
+
+	"github.com/cockroachdb/cockroach/gossip"
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/util/hlc"
+	"github.com/cockroachdb/cockroach/util/log"
+	"github.com/cockroachdb/cockroach/util/stop"
+)
+
+const (
+	// testTimeUntilStoreDead is the the test value for TimeUntilStoreDead to
+	// quickly mark stores as dead.
+	testTimeUntilStoreDead = 10 * time.Millisecond
+
+	// TestTimeUntilStoreDeadOff is the test value for TimeUntilStoreDead that
+	// prevents the store pool from marking stores as dead.
+	TestTimeUntilStoreDeadOff = 24 * time.Hour
+)
+
+type storeDetail struct {
+	desc            proto.StoreDescriptor
+	dead            bool
+	timesDied       int
+	foundDeadOn     time.Time
+	lastUpdatedTime time.Time // This is also the priority for the queue.
+	index           int       // index of the item in the heap, required for heap.Interface
+}
+
+// markDead sets the storeDetail to dead(inactive).
+func (sd *storeDetail) markDead(foundDeadOn time.Time) {
+	sd.dead = true
+	sd.foundDeadOn = foundDeadOn
+	sd.timesDied++
+	log.Warningf("store %s on node %s is now considered offline", sd.desc.StoreID, sd.desc.Node.NodeID)
+}
+
+// markAlive sets the storeDetail to alive(active) and saves the updated time
+// and descriptor.
+func (sd *storeDetail) markAlive(foundAliveOn time.Time, storeDesc proto.StoreDescriptor) {
+	sd.desc = storeDesc
+	sd.dead = false
+	sd.lastUpdatedTime = foundAliveOn
+}
+
+// storePoolPQ implements the heap.Interface (which includes  sort.Interface)
+// and holds storeDetail.
+type storePoolPQ []*storeDetail
+
+// Len implements the sort.Interface.
+func (pq storePoolPQ) Len() int {
+	return len(pq)
+}
+
+// Less implements the sort.Interface.
+func (pq storePoolPQ) Less(i, j int) bool {
+	return pq[i].lastUpdatedTime.UnixNano() < pq[j].lastUpdatedTime.UnixNano()
+}
+
+// Swap implements the sort.Interface.
+func (pq storePoolPQ) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+	pq[i].index, pq[j].index = i, j
+}
+
+// Push implements the heap.Interface.
+func (pq *storePoolPQ) Push(x interface{}) {
+	n := len(*pq)
+	item := x.(*storeDetail)
+	item.index = n
+	*pq = append(*pq, item)
+}
+
+// Pop implements the heap.Interface.
+func (pq *storePoolPQ) Pop() interface{} {
+	old := *pq
+	n := len(old)
+	item := old[n-1]
+	item.index = -1 // for safety
+	*pq = old[0 : n-1]
+	return item
+}
+
+// peek returns the next value in the priority queue without dequeuing it.
+// Lock is assumed held by caller.
```

Comment:
storePoolPQ doesn't have a mutex, so this reference is unclear. can you just say it's not thread safe?